toolchain/lower/handle.cpp

```diff
@@ -166,14 +167,75 @@ auto HandleBuiltin(FunctionContext& /*context*/, SemIR::InstId /*inst_id*/,
   CARBON_FATAL() << "TODO: Add support: " << inst;
 }
 
+static auto GetCalleeBuiltinFunctionKind(const SemIR::File& sem_ir,
+                                         SemIR::InstId callee_id)
+    -> SemIR::BuiltinFunctionKind {
+  if (auto bound_method =
+          sem_ir.insts().TryGetAs<SemIR::BoundMethod>(callee_id)) {
+    callee_id = bound_method->function_id;
+  }
+  callee_id = sem_ir.constant_values().Get(callee_id).inst_id();
+  if (!callee_id.is_valid()) {
+    return SemIR::BuiltinFunctionKind::None;
+  }
+  if (auto callee = sem_ir.insts().TryGetAs<SemIR::FunctionDecl>(callee_id)) {
+    const auto& function = sem_ir.functions().Get(callee->function_id);
+    return function.builtin_kind;
+  }
+  return SemIR::BuiltinFunctionKind::None;
+}
+
+static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
+                              SemIR::BuiltinFunctionKind builtin_kind,
+                              llvm::ArrayRef<SemIR::InstId> arg_ids) -> void {
+  switch (builtin_kind) {
+    case SemIR::BuiltinFunctionKind::None:
+      CARBON_FATAL() << "No callee in function call.";
+
+    case SemIR::BuiltinFunctionKind::IntAdd: {
+      // TODO: Move type checking to the point where we make the call.
+      if (arg_ids.size() != 2) {
+        break;
+      }
+      auto lhs_type = context.sem_ir().insts().Get(arg_ids[0]).type_id();
+      auto rhs_type = context.sem_ir().insts().Get(arg_ids[1]).type_id();
+      auto result_type = context.sem_ir().insts().Get(inst_id).type_id();
+      if (lhs_type != rhs_type || lhs_type != result_type ||
+          context.sem_ir().types().GetInstId(lhs_type) !=
+              SemIR::InstId::BuiltinIntType) {
+        break;
+      }
+      constexpr bool SignedOverflowIsUB = false;
+      context.SetLocal(
+          inst_id, context.builder().CreateAdd(context.GetValue(arg_ids[0]),
+                                               context.GetValue(arg_ids[1]), "",
+                                               false, SignedOverflowIsUB));
```

Comment:
```suggestion
                                               context.GetValue(arg_ids[1]), /*Name=*/"",
                                               /*HasNUW=*/false, SignedOverflowIsUB));
```

Trying to figure out what the not-really-provided parameters are... not that I intuitively understand HasNUW/HasNSW as "has no unsigned/signed wrap", but at least now I know the parameter names are a little opaque. Makes me kind of wonder if the constant should just be "OverflowIsUB" and pass it into both, e.g.:

```suggestion
                                               context.GetValue(arg_ids[1]), /*Name=*/"",
                                               /*HasNUW=*/OverflowIsUB, /*HasNSW=*/OverflowIsUB));
```

But I assume here your thought was this only would work for signed integers, so the unsigned value doesn't matter?

Reply:
Added parameter comments for NSW / NUW, since they could easily be transposed. I've added a name of "add" for the add instruction, which should obviate the need for a parameter comment there.

Regardless of whether the addition is signed or unsigned, we never want to set `NUW` -- that indicates that we're performing unsigned arithmetic for which overflow is UB, which for us it never is.